stock_picking_group_by_partner_by_carrier/report/report_delivery_slip.py

```diff
@@ -0,0 +1,103 @@
+    _description = "Delivery Slip Report"
+
+    @api.model
+    def _get_remaining_to_deliver(self, picking):
+        """Return dictionaries encoding pending quantities to deliver
+
+        Returns a list of dictionaries per sales order, encoding the data
+        to be displayed at the end of the delivery slip, summarising for
+        each order the pending quantities for each of its lines.
+        """
+        stock_move = self.env["stock.move"]
+
+        sales_data = OrderedDict()
+        for sale in picking.group_id.sale_ids.sorted(key=lambda s: s.id):
+            order_name = sale.get_name_for_delivery_line()
+            for line in sale.order_line.filtered(
+                lambda l: not l.display_type and l.product_id.type != "service"
+            ):
+                move = stock_move.search(
+                    [("sale_line_id", "=", line.id), ("picking_id", "=", picking.id)],
+                    limit=1,
+                )
+                qty = 0
+                if picking.state == "done" or not move:
+                    # If the picking is done, or if the line is not in any move
+                    # of this picking, we rely only in the sales order.
+                    qty = line.product_uom_qty - line.qty_delivered
+                elif picking.state not in ("cancel", "done"):
+                    # Else, we consider the amount reserved in the move.
+                    qty = (
+                        line.product_uom_qty - line.qty_delivered - move.product_uom_qty
+                    )
+
+                uom = move.product_uom if move else line.product_uom
+                uom_rounding = uom.rounding
+                if not float_is_zero(qty, precision_rounding=uom_rounding):
+                    if order_name not in sales_data:
+                        sales_data[order_name] = [
+                            {"is_header": True, "concept": order_name}
+                        ]
+                    sales_data[order_name].append(
+                        {
+                            "is_header": False,
+                            "concept": line.product_id.name_get()[0][-1],
+                            "qty": float_round(qty, precision_rounding=uom_rounding),
+                            "product": line.product_id,
+                            "uom": uom,
+                            "sale_order_line": line,
+                        }
+                    )
+
+        return sales_data
+
+    @api.model
+    def get_remaining_to_deliver(self, picking):
+        sales_data = self._get_remaining_to_deliver(picking)
+
+        # Check: being sales_data an *ordered* dictonary, maybe .values()
+        #        returns the items in orders, so would be as easy as:
+        #        `return sales_data.values()`
+        remaining_to_deliver = []
+        for _, sale_data in sales_data.items():
+            # Remove those orders having just the line of the title.
+            if len(sale_data) > 1:
```

Comment:
can't you determine this in `_get_remaining_to_deliver`?

Reply:
Yes, but is done this way on purpose: ```_get_remaining_to_deliver``` returns a list of lines grouped by sale order, while ```get_remaining_to_deliver``` flattens them. For example, the former returns ```d = {'so1': [line1, line2], 'so2': [line3, line4]}``` and the latter receives d and returns ```[['so1', 'line1', 'line2', 'so2', 'line3', 'line4']]```. So ```_get_remaining_to_deliver``` doesn't return the flatted structure to allow other modules (as we do) to change the base grouping based on other criteria.